### PR TITLE
feat(#338): Support on the `getUserBy` method which enables to talk to ldap to pull data if data is not available

### PR DESCRIPTION
**Pull request by riginoommen**
# Resolves 338

### 

# Explain the feature
Included the helper function for the ldap and user bridge which enable to pull data from ldap and create it in database by using the existing LDAP helper function

## Does this PR introduce a breaking change
No

## Screenshots
N/A
### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Does the change have appropriate unit tests?
- [x] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?

### CODE REVIEW

File: packages/user-service/src/resolver.ts [hunk 1]
```diff
@@ -11,7 +11,13 @@ export const UserGroupResolver = {
       const query: any = {};
       query[ keyParam ] = valueParam;
       return User.find( query )
-        .then( data => data )
+        .then( data => {
+          if ( (keyParam===`uid` || keyParam ===`rhatUUID`)&& _.isEmpty( data ) ) {
```
Comment (sayak-sarkar) [inline packages/user-service/src/resolver.ts:15]:
```suggestion
          if ( (keyParam ===`uid` || keyParam ===`rhatUUID`) && _.isEmpty( data ) ) {
```

File: packages/user-service/src/resolver.ts [hunk 2]
```diff
@@ -11,7 +11,13 @@ export const UserGroupResolver = {
...
+            return UserGroupAPIHelper.addUserLDAP( `${ keyParam }=${ valueParam }` ).then( ( user: any ) => [ user ] );
```
Comment (sayak-sarkar) [inline packages/user-service/src/resolver.ts:16]:
```suggestion
            return UserGroupAPIHelper.addUserLDAP( `${ keyParam } = ${ valueParam }` ).then( ( user: any ) => [ user ] );
```

**Comment by sayak-sarkar** [comment]
@riginoommen Please strictly follow the norms for PR naming.

File: packages/user-service/src/helpers.ts [hunk 3]
```diff
@@ -35,7 +38,34 @@ class UserGroupApiHelper {
...
+            title: response.title,
+            memberOf: groups,
+            isActive: true,
+            apiRole: ( groups.includes( 'one-portal-devel' ) ) ? `ADMIN` : 'USER',
+            createdBy: response.rhatUUID,
+            createdOn: moment.utc( new Date() )
+          } );
+          resolve( newUser.save() );
+        } else {
+          resolve( {} );
```
Comment (deshmukhmayur) [inline packages/user-service/src/helpers.ts:64]: If the response is empty, shouldn't it reject the Promise?
Instead of resolving with an empty object.

Comment (riginoommen) [reply packages/user-service/src/helpers.ts]: Updated

File: packages/user-service/src/helpers.ts [hunk 4]
```diff
@@ -61,7 +61,7 @@ class UserGroupApiHelper {
           } );
           resolve( newUser.save() );
         } else {
-          resolve( {} );
+          reject();
```
Comment (deshmukhmayur) [inline packages/user-service/src/helpers.ts:64]: Add some helpful error. Promises shouldn't just terminate without a reason.

> Like throw in plain old JavaScript, it's customary, but not required, to reject a promise with an Error object. The benefit of Error objects is they capture a stack trace, making debugging tools more helpful.
>
> – https://web.dev/promises/#promises-arrive-in-javascript!

File: packages/user-service/src/helpers.ts [hunk 5]
```diff
@@ -25,17 +28,44 @@ class UserGroupApiHelper {
...
-          console.error(`LDAP error: ` + error.message);
-        });
-        response.on(`end`, (result: any) => {
-          resolve(profile);
-        });
-      });
-    });
+        } );
+        response.on( `error`, ( error: Error ) => {
+          console.error( `LDAP error: ` + error.message );
```
Comment (deshmukhmayur) [inline packages/user-service/src/helpers.ts:33]: The promise should be rejected in `onError`

File: packages/user-service/src/helpers.ts (hunk 4 above)
Comment (riginoommen) [reply packages/user-service/src/helpers.ts:64]: actually its not an error it should send null response thats why i used resolve with null object.

Comment (riginoommen) [reply packages/user-service/src/helpers.ts]: Updated

File: packages/user-service/src/helpers.ts (hunk 5 above)
Comment (riginoommen) [reply packages/user-service/src/helpers.ts:33]: Updated

**Pull request merged**